lib/timers/promises.js

```diff
@@ -122,7 +124,110 @@ function setImmediate(value, options = {}) {
   });
 }
 
+function setInterval(after, value, options = {}) {
+  const args = value !== undefined ? [value] : value;
+  if (options == null || typeof options !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options',
+      'Object',
+      options);
+  }
+  const { signal, ref = true } = options;
+  if (signal !== undefined &&
+    (signal === null ||
+      typeof signal !== 'object' ||
+      !('aborted' in signal))) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options.signal',
+      'AbortSignal',
+      signal);
+  }
+  if (typeof ref !== 'boolean') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options.ref',
+      'boolean',
+      ref);
+  }
+  return {
+    [Symbol.asyncIterator]() {
+      let timeout,
+        deferred;
+      let active = true;
+      const iterator = {
+        next() {
+          if (!active) {
+            return PromiseResolve({
+              done: true,
+              value: undefined
+            });
+          }
+          // TODO(@jasnell): If a decision is made that this cannot
+          // be backported to 12.x, then this can be converted to
+          // use optional chaining to simplify the check.
+          if (signal && signal.aborted) {
+            return PromiseReject(
+              lazyDOMException('The operation was aborted', 'AbortError')
+            );
+          }
+          const promise = new Promise(
+            (resolve, reject) => {
+              deferred = { resolve, reject };
+            }
+          );
+          timeout = new Timeout((value) => {
+            if (deferred) {
+              deferred.resolve({
+                done: false,
+                value
+              });
+              deferred = undefined;
+            }
+          }, after, args, false, true);
+          if (!ref) timeout.unref();
+          insert(timeout, timeout._idleTimeout);
+          return promise;
+        },
+        return() {
+          active = false;
+          if (timeout) {
+            // eslint-disable-next-line no-undef
+            clearTimeout(timeout);
+          }
+          if (deferred) {
+            deferred.resolve({
+              done: true,
+              value: undefined
+            });
+            deferred = undefined;
+          }
+          if (signal) {
+            signal.removeEventListener('abort', onAbort);
+          }
+          return PromiseResolve({
+            done: true,
+            value: undefined
+          });
+        }
+      };
+      if (signal) {
+        signal.addEventListener('abort', onAbort, { once: true });
+      }
+      return iterator;
+
+      function onAbort() {
+        if (deferred) {
+          deferred.reject(
+            lazyDOMException('The operation was aborted', 'AbortError')
```

Comment:
Async iterators don't need to reject on cancellation or expose the abort error - it's enough to `.return` on the iterator which would cause whatever is wrapping it (for await, stream etc) to close.

Reply:
This is definitely a question I hoped to come up in review. 

@/jasnell had suggested `endOnSignal` to be an option to either reject or resolve for this.

Reply:
cc @jasnell - i think the mental model is that streams and iterators have a concept of "abort" built in unlike promises - so it's better to use it probably :]

Reply:
I can see it both ways really. An option makes it more complicated, but gives users the flexibility. I'm honestly not sure which should be the default here.

Reply:
I avoid doing cancellation "the regular JS way" with `.return` and not with a rejection/error unless we have a very compelling reason _not_ to do the "standard" thing.

Note that if we implement this with an async generator this solution will work "automatically" by putting cleanup in a `finally` block